### DISCUSSION TITLE
Support high resolution graphics / artwork upscale

### DISCUSSION THREAD

**Discussion by alula** [SuperTux/supertux#2049] [Ideas]
The issue is basically as the title suggests - SuperTux right now looks somewhat blurry when the magnification isn't set to 100% and the engine doesn't even have support for HD graphics.

I did some research on that topic and evaluated various automatic solutions for artwork upscaling - so far [waifu2x](https://github.com/nagadomi/waifu2x) yielded most promising results:
<details>
<summary>Upscale results (zero manual edits)</summary>

Command line used: `waifu2x-ncnn-vulkan -i $f -o upscaled/$f -n -1 -s 2`

![image](https://user-images.githubusercontent.com/6276139/147846802-b75b4e0a-a96a-4d3a-aba6-26222a6ef927.png)

![convex](https://user-images.githubusercontent.com/6276139/147846813-9f26a2a2-727c-40ed-96ca-bc116c925802.png)
![convex](https://user-images.githubusercontent.com/6276139/147846810-b1d7060c-eb86-4426-a1b0-428d9311c361.png)

![foresttiles-1](https://user-images.githubusercontent.com/6276139/147846821-4d11a3aa-1458-40e2-93c7-39c742ef718e.png)
![foresttiles-1](https://user-images.githubusercontent.com/6276139/147846823-6fba8d30-d405-401e-9abc-a231f54b047f.png)

![base](https://user-images.githubusercontent.com/6276139/147846847-4c34cddb-a1ec-4ba2-a957-5645b9b03f00.png)
![base](https://user-images.githubusercontent.com/6276139/147846852-18fda42a-d62f-4eb5-8231-a01e06fe88ad.png)

</details>

<details>
<summary>Applied in-game, quickly hacked together (I plan to push a PoC on my fork later):</summary>

![image](https://user-images.githubusercontent.com/6276139/147847069-0c312461-ba1b-4d01-9686-9c248fdd1613.png)
![image](https://user-images.githubusercontent.com/6276139/147847040-d94d94fa-1dc2-440e-bf99-510a5e5fca54.png)

![image](https://user-images.githubusercontent.com/6276139/147847086-5b7712e4-9637-4862-a145-c52d768f29a0.png)
![image](https://user-images.githubusercontent.com/6276139/147847094-3d4b6f23-bddc-4759-8946-1e620e8c7cf6.png)

![image](https://user-images.githubusercontent.com/6276139/147847111-f1252f93-6c32-431a-8dec-2809b3651376.png)
![image](https://user-images.githubusercontent.com/6276139/147847105-77ba0230-3f66-44b7-a6d5-1886fb66548d.png)

</details>

Obviously, it's not 100% perfect and needs some minor manual edits - notably by fixing the seams, however it's WAY less effort than redrawing the sprites from scratch.
The results seem to mostly depend on quality of source graphics - the tool prefers hand-drawn, non-blurry graphics - any blurriness or mix of hand drawn and pixel perfect lines drastically reduce the quality.

**Comment by HybridDog** [comment]
> I did some research on that topic and evaluated various automatic solutions for artwork upscaling

Which ones have you tried? Does it give better results than RealSR and Real-ESRGAN?

**Reply by alula** [reply]
I think that Real-ESRGAN is the best bet. Needs some extra tools to deal with seams and transparent edges and manual editing for perfect results.
